File: TP 2/Q02/Jogador.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// Definição da estrutura Jogador
struct Jogador {
    int id;
    int altura;
    int peso;
    int anoNascimento;
    char nome[100];
    char universidade[100];
    char cidadeNascimento[100];
    char estadoNascimento[100];
};

// Protótipos das funções para a estrutura Jogador
void setId(struct Jogador *jogador, int id);
int getId(struct Jogador jogador);
void setAltura(struct Jogador *jogador, int altura);
int getAltura(struct Jogador jogador);
void setPeso(struct Jogador *jogador, int peso);
int getPeso(struct Jogador jogador);
void setAnoNascimento(struct Jogador *jogador, int anoNascimento);
int getAnoNascimento(struct Jogador jogador);
void setNome(struct Jogador *jogador, const char *nome);
const char* getNome(struct Jogador jogador);
void setUniversidade(struct Jogador *jogador, const char *universidade);
const char* getUniversidade(struct Jogador jogador);
void setCidadeNascimento(struct Jogador *jogador, const char *cidadeNascimento);
const char* getCidadeNascimento(struct Jogador jogador);
void setEstadoNascimento(struct Jogador *jogador, const char *estadoNascimento);
const char* getEstadoNascimento(struct Jogador jogador);
struct Jogador clone(struct Jogador jogador);
void imprimir(struct Jogador jogador);
void ler(struct Jogador *jogador, int id);
void separarLinha(char *linha, char partes[8][100]);
void preencherAtributos(struct Jogador *jogador, char partes[8][100]);

// Implementação dos métodos para a estrutura Jogador
void setId(struct Jogador *jogador, int id) {
    jogador->id = id;
}

int getId(struct Jogador jogador) {
    return jogador.id;
}

void setAltura(struct Jogador *jogador, int altura) {
    jogador->altura = altura;
}

int getAltura(struct Jogador jogador) {
    return jogador.altura;
}

void setPeso(struct Jogador *jogador, int peso) {
    jogador->peso = peso;
}

int getPeso(struct Jogador jogador) {
    return jogador.peso;
}

void setAnoNascimento(struct Jogador *jogador, int anoNascimento) {
    jogador->anoNascimento = anoNascimento;
}

int getAnoNascimento(struct Jogador jogador) {
    return jogador.anoNascimento;
}

void setNome(struct Jogador *jogador, const char *nome) {
    strncpy(jogador->nome, nome, sizeof(jogador->nome));
    jogador->nome[sizeof(jogador->nome) - 1] = '\0';
}

const char* getNome(struct Jogador jogador) {
    return jogador.nome;
}

void setUniversidade(struct Jogador *jogador, const char *universidade) {
    strncpy(jogador->universidade, universidade, sizeof(jogador->universidade));
    jogador->universidade[sizeof(jogador->universidade) - 1] = '\0';
}

const char* getUniversidade(struct Jogador jogador) {
    return jogador.universidade;
}

void setCidadeNascimento(struct Jogador *jogador, const char *cidadeNascimento) {
    strncpy(jogador->cidadeNascimento, cidadeNascimento, sizeof(jogador->cidadeNascimento));
    jogador->cidadeNascimento[sizeof(jogador->cidadeNascimento) - 1] = '\0';
}

const char* getCidadeNascimento(struct Jogador jogador) {
    return jogador.cidadeNascimento;
}

void setEstadoNascimento(struct Jogador *jogador, const char *estadoNascimento) {
    strncpy(jogador->estadoNascimento, estadoNascimento, sizeof(jogador->estadoNascimento));
    jogador->estadoNascimento[sizeof(jogador->estadoNascimento) - 1] = '\0';
}

const char* getEstadoNascimento(struct Jogador jogador) {
    return jogador.estadoNascimento;
}

struct Jogador clone(struct Jogador jogador) {
    struct Jogador jogadorClone;
    jogadorClone.id = jogador.id;
    jogadorClone.altura = jogador.altura;
    jogadorClone.peso = jogador.peso;
    jogadorClone.anoNascimento = jogador.anoNascimento;
    setNome(&jogadorClone, getNome(jogador));
    setUniversidade(&jogadorClone, getUniversidade(jogador));
    setCidadeNascimento(&jogadorClone, getCidadeNascimento(jogador));
    setEstadoNascimento(&jogadorClone, getEstadoNascimento(jogador));
    return jogadorClone;
}

void imprimir(struct Jogador jogador) {
    printf("[%d ## %s ## %d ## %d ## %d ## %s ## %s ## %s]\n", jogador.id, jogador.nome, jogador.altura, jogador.peso, jogador.anoNascimento, jogador.universidade, jogador.cidadeNascimento, jogador.estadoNascimento);
}

void ler(struct Jogador *jogador, int id) {
    FILE *arquivo = fopen("/tmp/players.csv", "r");

    if (arquivo == NULL) {
        printf("Arquivo players.csv não encontrado.\n");
        return;
    }

    char linha[256];
    fgets(linha, sizeof(linha), arquivo);

    while (fgets(linha, sizeof(linha), arquivo) != NULL) {
        int jogadorId;
        char nome[100], universidade[100], cidadeNascimento[100], estadoNascimento[100];
        int altura, peso, anoNascimento;

        int numCampos = sscanf(linha, "%d,%99[^,],%d,%d,%99[^,],%d,%99[^,],%99[^\n]", &jogadorId, nome, &altura, &peso, universidade, &anoNascimento, cidadeNascimento, estadoNascimento);

        // Verifique se há campos vazios e substitua por "nao informado"
        if (numCampos < 8) {
            if (strchr(nome, ',') == NULL) {
                strcpy(nome, "nao informado");
            }
            if (strchr(universidade, ',') == NULL) {
                strcpy(universidade, "nao informado");
            }
            if (strchr(cidadeNascimento, ',') == NULL) {
                strcpy(cidadeNascimento, "nao informado");
            }
            if (strchr(estadoNascimento, ',') == NULL) {
                strcpy(estadoNascimento, "nao informado");
            }
        }

        // Remova vírgula final, se existir
        int len = strlen(estadoNascimento);
        if (estadoNascimento[len - 1] == ',') {
            estadoNascimento[len - 1] = '\0';
        }

        if (jogadorId == id) {
            setId(jogador, jogadorId);
            setAltura(jogador, altura);
            setPeso(jogador, peso);
            setAnoNascimento(jogador, anoNascimento);
            setNome(jogador, nome);
            setUniversidade(jogador, universidade);
            setCidadeNascimento(jogador, cidadeNascimento);
            setEstadoNascimento(jogador, estadoNascimento);

            imprimir(*jogador);
            break;
        }
    }

    fclose(arquivo);
}

void separarLinha(char *linha, char partes[8][100]) {
    char *token = strtok(linha, ",");
    int i = 0;
    while (token != NULL) {
        strncpy(partes[i], token, 100);
        token = strtok(NULL, ",");
        i++;
    }
}

void preencherAtributos(struct Jogador *jogador, char partes[8][100]) {
    setId(jogador, atoi(partes[0]));
    setNome(jogador, partes[1]);
    setAltura(jogador, atoi(partes[2]));
    setPeso(jogador, atoi(partes[3]));
    setUniversidade(jogador, partes[4]);
    setAnoNascimento(jogador, atoi(partes[5]));
    setCidadeNascimento(jogador, partes[6]);
    setEstadoNascimento(jogador, partes[7]);
}

int main() {
    struct Jogador jogador;

    while (1) {
        char input[10];
        scanf("%s", input);

        if (strcmp(input, "FIM") == 0) {
            break;
        }

        int id = atoi(input);
        ler(&jogador, id);
    }

    return 0;
}

File: TP 2/Q05/Jogador.java
import java.util.*;
import java.io.*;

class Jogador implements Comparable<Jogador> 
{
    private static int totalComparacoes = 0;

    // Atributos privados
    private int id;
    private int altura;
    private int peso;
    private int anoNascimento;
    private String nome;
    private String universidade;
    private String cidadeNascimento;
    private String estadoNascimento;

    // Construtores get de todos os int
    public int getId ()
    {
        return id;
    }

    public int getAltura ()
    {
        return altura;
    }

    public int getPeso ()
    {
        return peso;
    }

    public int getAnoNascimento ()
    {
        return anoNascimento;
    }

    // Construtores get de todos os String
    public String getNome ()
    {
        return nome;
    }

    public String getUniversidade ()
    {
        return universidade;
    }

    public String getCidadeNascimento ()
    {
        return cidadeNascimento;
    }

    public String getEstadoNascimento ()
    {
        return estadoNascimento;
    }

    // Construtores set de todos os int
    public void setId (int id)
    {
        this.id = id;
    }

    public void setAltura (int altura)
    {
        this.altura = altura;
    }

    public void setPeso (int peso)
    {
        this.peso = peso;
    }

    public void setAnoNascimento (int anoNascimento)
    {
        this.anoNascimento = anoNascimento;
    }

    // Construtores set de todos os String
    public void setNome (String nome)
    {
        this.nome = nome;
    }

    public void setUniversidade (String universidade)
    {
        this.universidade = universidade;
    }

    public void setCidadeNascimento (String cidadeNascimento)
    {
        this.cidadeNascimento = cidadeNascimento;
    }

    public void setEstadoNascimento (String estadoNascimento)
    {
        this.estadoNascimento = estadoNascimento;
    }

    // Construtor clone
    public Jogador clone ()
    {
        Jogador jogadorClone = new Jogador();

        // Clonando todos os int
        jogadorClone.setId(this.id);
        jogadorClone.setAltura(this.altura);
        jogadorClone.setPeso(this.peso);
        jogadorClone.setAnoNascimento(this.anoNascimento);

        // Clonando todos os String
        jogadorClone.setNome(this.nome);
        jogadorClone.setUniversidade(this.universidade);
        jogadorClone.setCidadeNascimento(this.cidadeNascimento);
        jogadorClone.setEstadoNascimento(this.estadoNascimento);

        return jogadorClone;
    }

    // Construtor ler
    public void ler (int id)
    {
        try
        {
            File arquivo = new File("/tmp/players.csv");
            Scanner scanner = new Scanner(arquivo);

            // Ignorar a primeira linha
                if (scanner.hasNextLine()) 
                {
                    scanner.nextLine();
                }

            while (scanner.hasNextLine())
            {
                String linha = scanner.nextLine();
                //System.out.println("Linha do arquivo: " + linha);
                List<String> partesSeparadas = separarLinha(linha);

                int jogadorId = Integer.parseInt(partesSeparadas.get(0));

                if (jogadorId == id)
                {
                    preencherAtributos(partesSeparadas);
                    
                    break;
                }
            }

            scanner.close();
        }catch (FileNotFoundException e) 
        {
            System.err.println("Arquivo players.csv não encontrado.");
        }
    }

    // Construtor para separar a linha em partes
    public static List<String> separarLinha(String linha) 
    {
        String[] partes = linha.split(",");
        List<String> partesSeparadas = new ArrayList<>();

        for (String parte : partes) 
        {
            if (parte.isEmpty()) 
            {
                partesSeparadas.add("nao informado");
            } 
            
            else 
            {
                partesSeparadas.add(parte);
            }
        }

        return partesSeparadas;
    }

    // Construtor para atribuir valores aos atributos de acordo com a linha do arquivo
    public void preencherAtributos(List<String> partesSeparadas) 
    {
        this.id = -1;  // Valor padrão para o ID
        this.nome = "nao informado";  // Valor padrão para o nome
        this.altura = -1;  // Valor padrão para a altura
        this.peso = -1;  // Valor padrão para o peso
        this.universidade = "nao informado";  // Valor padrão para a universidade
        this.anoNascimento = -1;  // Valor padrão para o ano de nascimento
        this.cidadeNascimento = "nao informado";  // Valor padrão para a cidade de nascimento
        this.estadoNascimento = "nao informado";  // Valor padrão para o estado de nascimento

        if (partesSeparadas.size() >= 1)
        {
            this.id = Integer.parseInt(partesSeparadas.get(0));
        }

        if (partesSeparadas.size() >= 2)
        {
            this.nome = partesSeparadas.get(1);
        }

        if (partesSeparadas.size() >= 3)
        {
            this.altura = Integer.parseInt(partesSeparadas.get(2));
        }

        if (partesSeparadas.size() >= 4)
        {
            this.peso = Integer.parseInt(partesSeparadas.get(3));
        }

        if (partesSeparadas.size() >= 5)
        {
            this.universidade = partesSeparadas.get(4);
        }

        if (partesSeparadas.size() >= 6)
        {
            this.anoNascimento = Integer.parseInt(partesSeparadas.get(5));
        }

        if (partesSeparadas.size() >= 7)
        {
            this.cidadeNascimento = partesSeparadas.get(6);
        }

        if (partesSeparadas.size() >= 8)
        {
            this.estadoNascimento = partesSeparadas.get(7);
        }
    }

    @Override
    public int compareTo(Jogador outroJogador) {
        totalComparacoes++; // Incrementa o contador de comparações
        return this.nome.compareTo(outroJogador.nome);
    }

    public static void selectionSort(List<Jogador> jogadores) {
        int n = jogadores.size();
    
        for (int i = 0; i < n - 1; i++) {
            int minIndex = i;
            for (int j = i + 1; j < n; j++) {
                if (jogadores.get(j).compareTo(jogadores.get(minIndex)) < 0) {
                    minIndex = j;
                }
            }
            // Troque a posição dos jogadores
            Jogador temp = jogadores.get(i);
            jogadores.set(i, jogadores.get(minIndex));
            jogadores.set(minIndex, temp);
        }
    }

    public void imprimir() {
        System.out.println("[" + this.id + " ## " + this.nome + " ## " + this.altura + " ## " + this.peso + " ## " +
               this.anoNascimento + " ## " + this.universidade + " ## " + this.cidadeNascimento + " ## " +
               this.estadoNascimento + "]");
    }

    public static void main(String[] args) {
        long inicioExecucao = System.nanoTime();
        Scanner scanner = new Scanner(System.in);
        List<Jogador> jogadoresSelecionados = new ArrayList<>();

        while (true) {
            String entrada = scanner.nextLine();

            if (entrada.equals("FIM")) {
                break;
            }

            try {
                int jogadorId = Integer.parseInt(entrada);
                Jogador jogador = new Jogador();
                jogador.ler(jogadorId);
                jogadoresSelecionados.add(jogador);
            } catch (NumberFormatException e) {
                System.err.println("ID inválido: " + entrada);
            }
        }

        // Utilize o algoritmo de ordenação por seleção para ordenar jogadoresSelecionados em ordem alfabética
        selectionSort(jogadoresSelecionados);

        // Exiba os jogadores em ordem alfabética
        for (Jogador jogador : jogadoresSelecionados) {
            jogador.imprimir();
        }

        scanner.close();

        long tempoExecucao = (System.nanoTime() - inicioExecucao) / 1000000; // Converta para milissegundos
        criarArquivoLog(tempoExecucao);
    }

    public static void criarArquivoLog(long tempoExecucao) 
    {
        try (PrintWriter writer = new PrintWriter("770855_sequencial.txt")) 
        {
            String matricula = "770855";
            writer.printf("%s\t%dms\t%d%n", matricula, tempoExecucao, totalComparacoes);
        } 
        
        catch (IOException e) 
        {
            e.printStackTrace();
        }
    }
}